### PR TITLE
Cleanup Deno.open and Deno.writeFile

### PR DESCRIPTION
**Pull request by dubiousjim**
This is part of larger series of filesystem commits (#4017).

* We refactor the internal `checkOpenOptions` function in cli/js/files.ts to parallel the stucture it will have in some concurrent PRs (see #4569, which also compares to peer languages).

* We remove a stray doc comment that got left/merged into `cli/js/ops/fs/open.ts`. (Currently, we're only putting comments in `cli/js/lib.deno.*`.)

* File system operarations tend to fail in different circumstances on Unix and Windows, and to give different errors when they do. However, with only a little effort on our part, we can ensure that errors triggered by violation of a `createNew` option are always AlreadyExists on all platforms. We now do this (with minimal runtime cost) in `op_open`.

* We change the implementation of `writeFile` to use the `create` option of `open` and also expose the `createNew` option (again, see #4569). This has the effect that the `mode` applied by `writeFile` is now only for newly created files, and honors the process's umask. Also `writeFile` no longer requires `--allow-read` permissions. (See #4423 for discussion about whether/when this should be required by `create`/`createNew`.)

* In this PR, `writeFile` has a very simple version of the internal `checkOpenOptions` function, which one may be tempted to inline. I left it factored out because later PRs which we might consider expand would expand this function.

* We changed the `writeFile` tests to accommodate the new behavior described two bullet points ago. One of the relevant tests (which exercises both `writeFile` and `stat`) is in `cli/js/tests/stat_test.ts`.

* We also add tests for the new `createNew` functionality on `writeFile`.

* Now that we have a `makeTempDir`, one `writeFile` test no longer needs to assume that the test machine has no file at `/baddir`.

* We also add tests for when one tries to `writeFile` to a path where a directory already exists. This gives different errors on Windows and Unix, but we verify that when `createNew` is true, the error is always AlreadyExists.

* We haven't yet implemented creating symlinks on Windows. On Unix, we verify that `writeFile` gives an error when writing to a path that contains any symlink when `createNew` is true, and that the error is always AlreadyExists. Also that it gives a (different) error when `createNew` is false, but writing to a path that contains a symlink to a directory. Also that it succeeds (gives no error) when `createNew` is false, and writing to a path that contains a symlink to a file, or a dangling symlink. The symlinks are left in place and we write at their target location. (These behaviors aren't ones we specifically design for; they're just what follows out of the implementations we're currently using. It seems good to test for them so that we'll be alerted if they ever change.)

### CODE REVIEW

File: cli/js/lib.deno.ns.d.ts [hunk 1]
```diff
@@ -1572,10 +1572,16 @@ declare namespace Deno {
     /** Defaults to `false`. If set to `true`, will append to a file instead of
      * overwriting previous contents. */
     append?: boolean;
+    /** Defaults to `false`. If set to `true`, no file, directory, or symlink is
+     * allowed to exist at the target location. When createNew is set to `true`,
+     * create is ignored. */
```
Comment (ry) [inline cli/js/lib.deno.ns.d.ts:1577]: Copy the doc string from Rust - you left out the bit about it being atomic. 

What error is thrown if a file already exists?

Comment (dubiousjim) [reply cli/js/lib.deno.ns.d.ts:1577]: In the PR description: "we can ensure that errors triggered by violation of a createNew option are always AlreadyExists on all platforms. We now do this...in op_open."

Am happy to include more of this info in the doc comments. But I'll wait for more feedback about this (and its sister PRs, pushing shortly) and will return to the doc comments at the end.

**Comment by dubiousjim** [comment]
Some of the `copyFile` tests in our codebase rely on the older behavior of `writeFile` wrt `mode`, and need to be updated after this PR lands. Whichever of this PR or #4584 is merged first, I can push a commit to the other PR that updates those `copyFile` tests.

File: cli/js/lib.deno.ns.d.ts (hunk 1 above)
Comment (dubiousjim) [reply cli/js/lib.deno.ns.d.ts]: Updated the doc-comment for createNew here (and on the other PRs too).

**Comment by bartlomieju** [comment]
@dubiousjim thank you for PR. Since you opened it there's been a lot of changes (there is no `mode` for `Deno.open()` anymore; symlink works on Windows). There are also a lot of conflicts. I'm gonna close this PR without merge, but feel free to open new one if there are changes that are still needed.

**Pull request closed without merging**